WooCommerce: Add SKU editor to the product form

justinshreve opened the pull request:
This PR introduces a generalized `FormClickToEditInput` component that behaves as described in [product mocks](https://www.dropbox.com/s/5gspfg0886qh3yw/add%20product.sketch?dl=0). It displays its value as text with a pencil/edit icon that transforms into an input when clicked. The value is set when the user clicks a checkmark.

This PR also hooks up the component to the product form and generates a SKU when entering a name.

Default state:
<img width="358" alt="screen shot 2017-06-01 at 2 19 17 pm" src="https://cloud.githubusercontent.com/assets/689165/26704696/d37af268-46e5-11e7-8f5c-3cb0f18a5be8.png">

SKU after typing a name:
<img width="358" alt="screen shot 2017-06-01 at 2 18 57 pm" src="https://cloud.githubusercontent.com/assets/689165/26704699/d94c3e40-46e5-11e7-8c60-11b8493c19f9.png">

SKU editing mode:
<img width="360" alt="screen shot 2017-06-01 at 2 19 06 pm" src="https://cloud.githubusercontent.com/assets/689165/26704703/de1e6074-46e5-11e7-8e87-f1001e4704db.png">

To Test:
* Visit `http://calypso.localhost:3000/store/products/:site/add`
* Type in a product name, and watch the SKU start appearing. A pencil icon should also appear.
* Click the pencil icon and see an input appear.
* Enter a new/custom SKU and click the checkmark.
* The SKU should update.

Closes #13653
cc @kellychoffman @jameskoster since this contains UI.

kellychoffman commented:
Nice. Could you add a class for when there's a value and when there isn't? This would be so I can add styles to each so that the spacing doesn't jump around:

![jumpy](https://cloud.githubusercontent.com/assets/5835847/26708886/9871e744-471c-11e7-9917-4601cdab5c0c.gif)

Open to other ways of how to address this.

justinshreve commented:
@kellychoffman Done.

kellychoffman commented:
👍 LGTM!

ryelle commented on client/extensions/woocommerce/app/products/product-form-details-card.js:
```diff
@@ -89,10 +114,20 @@ export default class ProductFormDetailsCard extends Component {
 							<FormTextInput name="name" value={ product.name || '' } onChange={ this.setName } />
 						</FormFieldSet>
+						<FormFieldSet className="products__product-form-details-basic-sku">
+							<FormLabel htmlFor="sku">{ __( 'SKU:' ) }</FormLabel>
```
The sku input needs to have the id `sku` if you're using `htmlFor` here — the `for` attribute matches to the `id` of the input, not the name (the same is true of the other labels on this page). Right now this isn't doing anything — and you can check this by clicking on the label text, if it's correctly linked, i'll put focus into the input.

ryelle commented on client/extensions/woocommerce/components/form-click-to-edit-input/index.js:
```diff
@@ -0,0 +1,119 @@
+					autoFocus
+					className="form-click-to-edit-input__input"
+				/>
+				<Button
```
This `Button` could use `aria-label={ translate( 'Update SKU' ) }`, to help future accessibility efforts.

ryelle commented on client/extensions/woocommerce/components/form-click-to-edit-input/index.js:
```diff
@@ -0,0 +1,119 @@
+				</span>
+
+				{ ! disabled && (
+					<Button
```
Same as the update button, this could use an `aria-label`.

ryelle commented on client/extensions/woocommerce/components/form-click-to-edit-input/index.js:
```diff
@@ -0,0 +1,119 @@
+		} );
+		return (
+			<span className={ classes }>
+				<span className="form-click-to-edit-input__text">
```
Maybe you can put the `onClick` on here too? I clicked the text twice during testing thinking it would open the edit box

ryelle commented on client/extensions/woocommerce/app/products/product-form-details-card.js:
```diff
@@ -29,6 +31,11 @@ export default class ProductFormDetailsCard extends Component {
 
+		const { product } = props;
+		this.state = {
+			updateSkuOnNameChange: ! product.name ? true : false,
```
Why not just `updateSkuOnNameChange: ! product.name`? The logic here is that the SKU is not automatically updated when an existing product (one with a name) is being updated, right?

ryelle commented on client/extensions/woocommerce/app/products/product-form-details-card.js:
```diff
@@ -38,7 +45,24 @@ export default class ProductFormDetailsCard extends Component {
+
+	setSku = ( sku ) => {
+		const { product, editProduct } = this.props;
+		editProduct( product, { sku } );
```
The `sku` isn't converted to a slug format like it is when generated in the title, is that ok?

justinshreve replied:
Yeah that's OK. WooCommerce can handle other SKUs formatted in different ways. The default ones we are generating based off name will be slug like format.

justinshreve commented:
@ryelle Thanks for the review. I made some fixes: Simplified the product.name check, correctly use IDs on our inputs, added aria labels for the edit and save icons, made the text editable by clicking, and added an onBlur.

ryelle commented on client/extensions/woocommerce/components/form-click-to-edit-input/index.js:
```diff
@@ -0,0 +1,127 @@
+		const props = { ...this.props,
+			onChange: this.onInputChange,
+			value: this.state.value,
+			updateAriaLabel: null,
```
You could move this into defaultProps

ryelle commented on client/extensions/woocommerce/components/form-click-to-edit-input/index.js:
```diff
@@ -0,0 +1,127 @@
+		return (
+			<span className="form-click-to-edit-input__wrapper editing">
+				<FormTextInput
+					{ ...props }
```
Now I'm getting the following error in console:

> Unknown props `updateAriaLabel`, `editAriaLabel` on <input> tag. Remove these props from the element.

You could omit the labels before `...props`

justinshreve commented:
@ryelle Thanks, done. Amended to the previous commit.
